src/coreclr/vm/gcstress.h

```diff
@@ -72,40 +72,34 @@ namespace GCStressPolicy
     private:
         // This static controls whether GC stress may induce GCs. EEConfig::GetGCStressLevel() still
         // controls when GCs may occur.
-        static Volatile<DWORD> s_nGcStressDisabled;
-
-        bool m_bAcquired;
+        static int s_nGcStressDisabled;
 
     public:
         InhibitHolder()
-        { LIMITED_METHOD_CONTRACT; ++s_nGcStressDisabled; m_bAcquired = true; }
+        { LIMITED_METHOD_CONTRACT; GlobalDisable(); }
 
         ~InhibitHolder()
         { LIMITED_METHOD_CONTRACT; Release(); }
 
         void Release()
-        {
-            LIMITED_METHOD_CONTRACT;
-            if (m_bAcquired)
-            {
-                --s_nGcStressDisabled;
-                m_bAcquired = false;
-            }
-        }
+        { LIMITED_METHOD_CONTRACT; GlobalEnable();}
 
         friend bool IsEnabled();
         friend void GlobalDisable();
         friend void GlobalEnable();
     } UNUSED_ATTR;
 
     FORCEINLINE bool IsEnabled()
-    { return InhibitHolder::s_nGcStressDisabled == 0U; }
+    { return Interlocked::ExchangeAdd(&InhibitHolder::s_nGcStressDisabled, 0) == 0; }
```

Comment:
Is the barrier on read actually necessary for something; or is just a defensive measure?

Reply:
just a defensive measure - in case it is important that we see the flag no earlier/later than we do something.

We had a volatile read, which could be arbitrarily prefetched. So I figured - either this should be an ordinary read or full-fence. 

It probably does not matter though. Nothing should strictly depend on the order, thus there will always be races. Just need to make sure the read is not optimized away.

Reply:
Maybe it should be just compiler fence, so it is less confusing, why interlocked.